Headless docs improvements (#79)
ilastik/ilastik.github.io: documentation/basics/headless.md

When running ilastik in headless mode, you must specify at least:

- The `--headless` flag, to trigger headless operation;
- Your pre-trained ilastik `.ilp` project file, via the `--project` argument (e.g. `--project=MyPixelClassProj.ilp`);
- One or more files to be processed as a batch. Those should always be given as the last arguments to the command line.

For example, this command will run your classifier over 2 additional images (here we assume your classifier was trained on 2D images):

Linux:

    $ cd ilastik-1.3.2-Linux/
    $ ./run_ilastik.sh --headless --project=MyProject.ilp my_next_image1.png my_next_image2.png

Mac:

    $ ./ilastik-1.3.2-OSX.app/Contents/ilastik-release/run_ilastik.sh  --headless --project=MyProject.ilp my_next_image1.png my_next_image2.png

Windows:

    $ cd "\Program Files\ilastik-1.3.2"
    $ .\ilastik.bat --headless --project=MyProject.ilp my_next_image1.png my_next_image2.png

**Note:** the following examples use linux shell syntax, but the options are the same for all platforms.

### Important Notes:

> - Except for options passed with the `--export_source` argument (see below), don't specify any values with space-characters in them, even if inside quotes (e.g. `"my file name.h5"`, or `"(10, 20)"`). Due to a [bug in cpython](https://bugs.python.org/issue22433) those might be misinterpreted.

Reviewer: Could you please test if it works with an export source containing a space and and image as a positional argument? I would expect that you'd have to specify the type of data as an optional argument with the image in order to get it to work.

- Except for options passed with the `--export_source` argument (see below), don't specify any values with space-characters in them, even if inside quotes (e.g. `"my file name.h5"`, or `"(10, 20)"`). Due to a [bug in cpython](https://bugs.python.org/issue22433) those might be misinterpreted.
- For paths to hdf5 datasests (either input or output), ilastik uses the same conventions as the generic [h5ls](https://support.hdfgroup.org/HDF5/Tutor/cmdtoolview.html#h5ls) utility. That is, the hdf5 dataset name should be appended to the file path: `/path/to/my_file.h5/internal/path/to/dataset`.
- In the current "headless" implementation of object classification, the entire image is loaded into RAM in one go, and then object classification is run on it.  Therefore, there is a limit to how large your input image can be.

> - ilastik's command line options use underlines rather than dashes to separate words (e.g.: `--cutout_subregion` and not `--cutout-subregion`). If you mix them up, you might get strange errors. See [Known Issues](#known-issues) for more info.

Reviewer 2: The _argparse_ module actually converts flags defined as `--cutout-subregion` into `args.cutout_subregion`, and I'm positive we can modify _ArgumentParser_ to accept either `-` or `_` as separators.

Author: Yup. We could do something like `parser.add_argument('--my_flag', '--my-flag', ...)` everywhere we create arguments, but we'd have to do that even for the arguments which are created dynamically and get that merged before the next release. And still there will be people stuck on older versions, so we might as well have that documented somewhere, I guess.

But I agree. Lets fix it =D

    E.g.: If your image is a png file with axis x y c, then you could run a headless prediction on a 50x50 square starting from 10x10 and containing all channels by specifying `--cutout_subregion="[(10,10,None),(60,60,None)]"`.
- `--export_dtype` The pixel type to convert your results to.  Choices are: `uint8`, `uint16`, `uint32`, `int8`, `int16`, `int32`, `float32`, `float64`.  Note that some formats don't support every pixel type.
- `--output_axis_order` Transpose the storage order of the results. For example, this affects the sliced dimension for stack outputs.
- `--pipeline_result_drange` Pipeline result data range (min,max) BEFORE normalization, e.g. `"(0.0,1.0)"`
- `--export_drange` Exported data range (min,max) AFTER normalization, e.g. `"(0,255)"`

[Data Export Applet]: {{site.baseurl}}/documentation/basics/export.html#data-export-applet-ss

## Headless Mode for Pixel Classification

When running the Pixel Classification Workflow in headless mode, the available options for the `--export_source` flag are:

> - `"Probabilities"`, which exports a multi-channel image where pixel values represent the probability that that pixel belongs to the label represented by that channel;

Reviewer: ```suggestion
- `"Probabilities"`, which exports a multi-channel image where pixel values represent the probability that that pixel belongs to the class represented by that channel;
```

> - `"Simple Segmentation"`, which produces an image where pixel values indicate the class to which a pixel belongs. For this image, every pixel with the same value should belong to the same class of pixels;

Reviewer: ```suggestion
- `"Simple Segmentation"`, which produces a single channel image of integer pixel values, where pixel values indicate the class to which a pixel belongs. For this image, every pixel with the same value should belong to the same class of pixels;
```

Reviewer: okay I see you use the term "image" synonymous for "single channel image" and indicate where images are multichannel. That's also good.

Author: "Explicit is better than implicit". Fixing it to single-channel

- `Labels`, which outputs an image representing the users' manually created annotations.

## Headless Mode for Object Classification

When running the Object Classification Workflow in headless mode, the available options for the `--export_source` flag are:
- `"Object Predictions"` (default when nothing is specified), which exports a label image of the object class predictions;
- `"Object Probabilities"`, which exports a multi-channel image volume of object prediction probabilities instead of a label image (one channel for each prediction class);
- `"Blockwise Object Predictions"` or `"Blockwise Object Probabilities"`, which is analogous to `"Object Predictions"` and `"Object Probabilities"` respectively, but the image will be processed in independent blocks, which is useful when your image won't fit in RAM. Note that the values for the block size and halo cannot be configured via command line parameters and must be set in yout `.ilp` project file via the graphical user interface. See [Blockwise Object Classification Applet]({{site.baseurl}}/documentation/objects/objects.html#preparing-for-large-scale-prediction---blockwise-object-classification-applet).

> - `"Pixel Probabilities"`, which exports the pixel prediction images of the pixel classification part of that workflow. Only valid if you specified an `.ilp` `--project` that was created with the  "Pixel Classification + Object Classification" workflow

Reviewer: missing punctuation mark at the end

[Object Classification Workflow]: {{site.baseurl}}/documentation/objects/objects.html

Depending on which variant of the [Object Classification Workflow] you used to create your `.ilp` project file, you may need to provide more than one input image for each volume of data you want to process (e.g. "Raw Data" and "Segmentation Image" for , or "Raw Data" and "Prediction Maps"). These input images correspond to the tabs in the `Input Data` applet of the Object Classification Workflow and must be provided on the command-line.  To specify which is which, prefix the list of input files with either `--raw_data`, `--segmentation_image`, or `--prediction_maps` accordingly. Here's an example of invoking ilastik with an Object Classification Workflow that takes a Segmentation Image as input:

    $ ./run_ilastik.sh --headless \
                       --project=MyObjClassFromPredictMap.ilp \
                       --table_filename=/tmp/exported_object_features.csv \
                       --export_source="Object Predictions" \
                       --raw_data "my_grayscale_stack_1/*.png" \

>                        --segmentation_image my_unclassified_objects_1.h5/binary_segmentation_volume

Reviewer: adding an example with, or at least mentioning `--prediction_maps` would be nice

- `Labels`, which outputs an image representing the users' manually created annotations.

## Headless Mode for Autocontext Workflow

The [Autocontext]({{site.baseurl}}/documentation/autocontext/autocontext) workflow is essentially two runs of pixel classification, with the latter one using the output of the first one as an additional input. When running the Autocontext Workflow in headless mode, the available options for the `--export_source` are analogous to those of the simple Pixel Classification Workflow, but you can export the results of any of the two pixel classification stages involved in this workflow. Here's the full list of options:

- `"Probabilities Stage 1"`;
- `"Probabilities Stage 2"`;
- `"Simple Segmentation Stage 1"`;
- `"Simple Segmentation Stage 2"`;
- `"Uncertainty Stage 1"`;
- `"Uncertainty Stage 2"`;
- `"Features Stage 1"`;
- `"Features Stage 2"`;
- `"Labels Stage 1"`;
- `"Labels Stage 2"`;
- `"Propabilities All Stages"`: equivalent to exporting both `"Probabilities Stage 1"` and `"Probabilities Stage 2"`;
- `"Input Stage 1"`: output your raw input image that was fed into the first stage of the workflow;

> - `"Input Stage 2"`, wich is the input received by the second Pixel Classification stage in the workflow

Reviewer: "wich" -> "which"

    
The RAM limit is not perfectly respected in all cases, so you may want to leave some buffer if your RAM budget is strict.

## Running your own Python scripts

For developers and power-users, you can run your own ilastik-dependent python scripts using the interpreter shipped within the ilastik install tree.  The interpreter is located in the `bin` directory:

    # Linux
    $ ./ilastik-1.1.7-Linux/bin/python -c "import ilastik; print ilastik.__version__"
    1.1.7
    
    # Mac
    $ ./ilastik-1.1.7-OSX.app/Contents/ilastik-release/bin/python -c "import ilastik; print ilastik.__version__"
    1.1.7

## Known Issues

ilastik's headless mode will sometimes throw exceptions and output a stacktrace instead of letting you known why your command line arguments are wrong. Though those issues are being worked on, here are some hints and workarounds you can use to get by:

> ### RuntimeError: Could not find one or more input files.  See logged errors.

Reviewer: ```suggestion
* 'RuntimeError: Could not find one or more input files.  See logged errors.': ...
```

It looks a bit odd with the heading
